docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
 nodes from that dataset. (Useful for a self-join, for example.)
 Note that producing two scan nodes like this will perform all
 reads and decodes twice.
+
+Constructing ``ExecNode`` using Options
+=======================================
+
+Using the execution plan we can construct various queries. 
+To construct such queries, we have provided a set of building blocks
+or referred as :class:`ExecNode` s. These nodes provide the ability to 
+construct operations like filtering, projection, join, etc. 
+
+This is the list of :class:`ExecutionNode` s exposed;
+
+1. :class:`SourceNode`
+2. :class:`FilterNode`
+3. :class:`ProjectNode`
+4. :class:`ScalarAggregateNode`
+5. :class:`SinkNode`
+6. :class:`ConsumingSinkNode`
+7. :struct:`OrderBySinkNode`
+8. SelectK-SinkNode
+9. Scan-Node
+10. :class:`HashJoinNode`
+11. Write-Node
+12. :class:`UnionNode`
+
+There are a set of :class:`ExecNode` s designed to provide various operations required
+in designing a streaming execution plan. 
+
+``SourceNode``
+--------------
+
+:struct:`arrow::compute::SourceNode` can be considered as an entry point to create a streaming execution plan. 
+A source node can be constructed as follows.
+
+:class:`arrow::compute::SoureNodeOptions` are used to create the :struct:`arrow::compute::SourceNode`. 
+The :class:`Schema` of the data passing through and a function to generate data 
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>` 
+are required to create this option::
+
+    // data generator
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen() { ... }
```

Comment:
I think that's ok then, but it might be nice to explain that this represents a generator, since it may not otherwise be clear that you should expect to call the std::function repeatedly.